Fix boost on mac and linux: correctly search for boost and set -mt suffix when needed.

bredelings opened the pull request:
Currently meson incorrect determines that the modname for boost program options is boost_program_options-mt instead of boost_program_options if the filename is `libboost_program_options-mt.dylib`.  This makes meson unable to detect boost installed from macports.

Fix this by correctly computing the modname from the filename, and factor out all the code that does this into one place instead of doing in a different way in each place.

Fixes #2823

ignatenkobrain commented on mesonbuild/dependencies/misc.py:
```diff
@@ -296,22 +303,11 @@ def detect_lib_modules_win(self):
-            if self.static:
-                modname = modname[3:]
-            self.lib_modules[modname] = fname
+            self.lib_modules[self.modname_from_filename(fname)] = fname
```
__[Flake8]__

[F821] undefined name 'fname'


<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/dependencies/misc.py:
```diff
@@ -296,22 +303,11 @@ def detect_lib_modules_win(self):
                 if self.static:
-                    modname = modname[3:]
-                    self.lib_modules[modname] = fname
+                    self.lib_modules[self.modname_from_filename(fname)] = fname
```
__[Flake8]__

[F821] undefined name 'fname'


<!-- Comment by [SideCI](https://sideci.com) -->

bredelings commented:
BTW, I replaced `(,fname) = os.path.split(entry)` with `fname = os.path.basename(entry)` since the latter seems more readable.  Is there some reason to prefer the former?

bredelings commented:
Fixes #2823

nirbheek commented:
@NickeZ please review :)

mojca commented:
@bredelings: I applied this patch on top of 0.44.0 and then tried to build [bali-phy](https://github.com/bredelings/Bali-Phy), but I'm still getting the "not found" error when using `-Dwith-system-boost=True`.

```
The Meson build system
Version: 0.44.0
Source dir: ...
Build dir: ...
Build type: native build
Project name: bali-phy
Native C compiler: /usr/bin/clang (clang 9.0.0)
Appending CFLAGS from environment: '-pipe -Os -arch x86_64'
Appending LDFLAGS from environment: '-L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64'
Appending CPPFLAGS from environment: '-I/opt/local/include'
Native C++ compiler: /usr/bin/clang++ (clang 9.0.0)
Appending CXXFLAGS from environment: '-pipe -Os -stdlib=libc++ -arch x86_64'
Appending LDFLAGS from environment: '-L/opt/local/lib -Wl,-headerpad_max_install_names -arch x86_64'
Appending CPPFLAGS from environment: '-I/opt/local/include'
Build machine cpu family: x86_64
Build machine cpu: x86_64
Has header "sys/resource.h": YES
Has header "fenv.h": YES
Checking for function "log2": YES
Configuring config.h using configuration
Found pkg-config: /opt/local/bin/pkg-config (0.29.2)
Native dependency cairo found: YES 1.14.12

Meson encountered an error in file meson.build, line 76, column 2:
Requested Boost library 'program_options' not found
```

NickeZ commented:
Overall looks good to me. As long as the tests pass I would merge it.

It is unfortunate that we don't have better tests since it is difficult to see if this changes the behavior in all the edge cases. Like for example, if I have `program_options`, `program_options-mt` and `program_options-mt-gd` installed. Does it still pick the correct library if dependency is called with `static=true` and/or `threaded=multi`.

bredelings commented:
The not-found error is a regression in meson handling of -mt, so 
interesting that you should mention that.  We could probably merge this 
after meson 0.44.1 is released, since I think it will contain the patch 
from here (https://github.com/mesonbuild/meson/issues/2823) and will (I 
suspect) be released in a week or so.

-BenRI


On 12/25/2017 04:29 AM, Niklas Claesson wrote:
>
> Overall looks good to me. As long as the tests pass I would merge it.
>
> It is unfortunate that we don't have better tests since it is 
> difficult to see if this changes the behavior in all the edge cases. 
> Like for example, if I have |program_options|, |program_options-mt| 
> and |program_options-mt-gd| installed. Does it still pick the correct 
> library if dependency is called with |static=true| and/or 
> |threaded=multi|.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_mesonbuild_meson_pull_2824-23issuecomment-2D353866318&d=DwMFaQ&c=imBPVzF25OnBgGmVOlcsiEgHoG1i6YHLR0Sj_gZ4adc&r=yDQgdhRIVlGWlo1dehDX4FDHdLhgFHI_XNBo6vRcIuw&m=KIBh1IAHnvvXTwfZBZYcQl32yIyWDCyt_hXgzO7XaOI&s=xoXMrRnXC1bi-FDSRMVtOA1zIilS0FsqFyp2yD6dKlc&e=>, 
> or mute the thread 
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAIj7OPpZ2NE449WKIg7gpw9O446QPv0ks5tD5U7gaJpZM4RLgf1&d=DwMFaQ&c=imBPVzF25OnBgGmVOlcsiEgHoG1i6YHLR0Sj_gZ4adc&r=yDQgdhRIVlGWlo1dehDX4FDHdLhgFHI_XNBo6vRcIuw&m=KIBh1IAHnvvXTwfZBZYcQl32yIyWDCyt_hXgzO7XaOI&s=ExknW2zpIcuhrH6yV3dkPghbYqOSjVxprY1s7dIPbw0&e=>.
>

bredelings commented:
We do have `test cases/frameworks/1 boost`.  Why did this not trigger in travis?

bredelings commented:
@mojca I'll check again later, but for now did you set BOOST_ROOT=/opt/local ?

bredelings commented:
For me, this does not work:
`CPPFLAGS=-I/opt/local/include LDFLAGS=-L/opt/local/lib meson ../master/ meson -Dwith-system-boost=True`
But this does work:
`BOOST_ROOT=/opt/local CPPFLAGS=-I/opt/local/include LDFLAGS=-L/opt/local/lib meson ../master/ meson -Dwith-system-boost=True`
It seems wrong to have to set an environment variable specifying the root for every library - is there some way to set this globally, like for cross builds?

Hmm... it seems that the way searching for boost works, it picks some specific dirs and checks if they contain files with names that match a pattern.  So, LDFLAGS has no effect on this search.

mojca commented:
Indeed, with `BOOST_ROOT=/opt/local` it (kind of) works:
```
Dependency Boost (program_options, random, system, filesystem, chrono) found: YES 1.65.1, /opt/local
```
That is: the configure goes through, but the linker later fails, being unable to find the library because the `-mt` suffix is missing:
```
$ /usr/bin/clang++  -o src/alignment-info 'src/alignment-info@exe/tools_alignment-info.cc.o' -L/opt/local/lib -pipe -Os -stdlib=libc++ -arch x86_64 -Wl,-headerpad_max_install_names -arch x86_64 src/libbali-phy.a -L/opt/local/lib -lboost_program_options -lboost_random -lboost_system -lboost_filesystem -lboost_chrono -L/opt/local/lib -lboost_program_options -lboost_random -lboost_system -lboost_filesystem -lboost_chrono
$ ld: library not found for -lboost_program_options
```
And I totally agree that having to manually specify the prefix for every single dependency feels super weird to me. Sure, it makes sense to specify it when I want to take a dependency from elsewhere, but from the perspective of a package manager where one knows in advance that all dependencies should come from one certain well-known prefix, it feels waste of time and very error-prone to have to specify it for every single dependency.

NickeZ commented:
@mojca Which OS are you using that puts boost in /opt/local? That sounds like an unusual place for a standard directory.

@bredelings Yes, I'm aware of the tests we have. But they only test the very simple case where boost is installed with the system provided package manager (linux) or the binaries provided from SF.net (windows). There are many other ways you can build and install boost which we don't test.

mojca commented:
I'm using macOS with the [MacPorts](https://github.com/macports/) package manager. The path `/opt/local` is default (but it could be any given path and I particularly don't like the path being hardcoded anywhere). The reason why the path might sound a tiny bit strange is that all the "usual" prefixes are reserved for binaries provided by Apple.

I believe it should be straightforward to write a test for this scenario. Just install a super simple hello world library with that name (`libboost_program_options-mt.dylib`) under any given prefix (it could be in `/tmp` or under the directory where the tests are being run) and try to find the library and link against it. No need to actually test a fully-functional boost installation. (Of course just as one of many possible tests.)

bredelings commented:
@mojca Would the link work if you _also_ supply LDFLAGS specifically?  The line I showed about (with CXXFLAGS and LDFLAGS) linked for me.

bredelings commented:
@NickeZ Sorry if I mis-spoke.  I'm just confused due to inexperience - why aren't the travis tests currently failing on osx?  Is it the case that travis doesn't run tests under `test cases`?

mojca commented:
@bredelings: The `LDFLAGS` have been set as can also be seen from the command being executed. The problem is in the
```
-lboost_program_options
```
flag which should have been
```
-lboost_program_options-mt
```
instead.

What's the output of `otool -L /opt/local/bin/bali-phy` on your machine?

NickeZ commented:
I wasn't aware that the tests we have weren't running on mac. I'm looking into it.

bredelings commented:
@mojca Good catch.  On my system, we detect the '-mt' form, but link against the non-mt form, and both exist. Let me look into this...

bredelings commented:
@NickeZ I think we might run the unit-tests, but not the `test cases` tests?

bredelings commented:
Argh.  The code in detect_modules can be consistent with either validate_requested or get_link_args( ), but not both.  Also, the get_link_args( ) routine calling compiler.find_library('boost_'+module_name).  If we are going to call compiler.find_library( ) ignore, we should call it in detect_modules( ). 

One other issues is that almost any change to the meson.build file results in 30 seconds of waiting, if we have a boost dependency.  Maybe this is because we are calling compiler.find_library( ) from get_link_args( )?

mojca commented:
[They are](https://travis-ci.org/mesonbuild/meson/jobs/320460631).

ignatenkobrain commented on mesonbuild/dependencies/misc.py:
```diff
@@ -334,19 +332,23 @@ def detect_lib_modules_nix(self):
+                else:
+                    continue
+                modname = self.modname_from_filename(entry)
+                if not modname in self.lib_modules:
```
__[Flake8]__

[E713] test for membership should be 'not in'


<!-- Comment by [SideCI](https://sideci.com) -->

bredelings commented:
@mojca @NickeZ The call to compiler.find_library( ) in get_link_args( ) was throwing away the '-mt' suffix, so I switched to just using the files we had already discovered in detect_lib_modules_nix( ).  This incidentally is much faster, because compiler.file_library( ) is very slow and get_link_args( ) is called a huge number of times.

jpakkane commented:
Tests with dependencies are not run on Mac or Windows by default because installing all the dependencies reliably is a pain and the build machines are stateless.

bredelings commented:
The travis build shows no errors except a timeone in one of the 6 jobs - can someone restart this?

bredelings commented:
@nirbheek @NickeZ I think a number of further improvements are possible.  These would include fixing cross builds, finding boost if its on in /usr/lib or /usr/local/lib, etc.
* Should these be discussed on the mailing list, or by opening new issues (or both)?
* Should I just push changes to this PR, or open new PRs?

NickeZ commented:
@bredelings You can probably improve the situation as much as possible for mac in this PR. But if you plan to implement cross-compilation I think that should go into a separate PR. This is not really my call anyway since I'm only a contributer.

@jpakkane Boost is installed with homebrew on the jenkins macos build images. Should we start testing against it?

jpakkane commented:
If the way the CI image ships it is representative with how people are using it in the real world, why not.

ignatenkobrain commented:
Flake8 detected 6 issues on efb45b251da4bce2f1d2ebd357356c768d3e0608
Visit https://sideci.com/gh/19784232/pull_requests/2824 to review the issues.


<!-- Comment by [SideCI](https://sideci.com) -->

bredelings commented:
After the last change we can find boost in the standard location on macports without needing to set BOOST_ROOT=/opt/local

bredelings commented:
@NickeZ Enabling checks on homebrew is a good idea.  I have remote access to a Mac, though it is not my own, so perhaps I can help.

ignatenkobrain commented on mesonbuild/dependencies/misc.py:
```diff
@@ -25,7 +25,7 @@
 from .. import mlog
 from .. import mesonlib
-from ..mesonlib import Popen_safe, extract_as_list
+from ..mesonlib import Popen_safe, extract_as_list, EnvironmentException
```
__[Flake8]__

[F401] '..mesonlib.Popen_safe' imported but unused


<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/dependencies/misc.py:
```diff
@@ -231,25 +264,30 @@ def validate_requested(self):
-            ifile = open(os.path.join(self.incdir, 'boost', 'version.hpp'))
-        except FileNotFoundError:
+            version = self.compiler.get_define('BOOST_LIB_VERSION', '#include <boost/version.hpp>', self.env, self.get_compile_args(), [])
+        except EnvironmentException:
```
__[Flake8]__

[F821] undefined name 'EnvironmentException'


<!-- Comment by [SideCI](https://sideci.com) -->

bredelings commented:
@ignatenkobrain I'm getting an error from ci/sideci because I am importing an exception that I only catch, but do not throw.  If I don't import it, I get another error for not defining it.  What should I do?

ignatenkobrain commented on mesonbuild/dependencies/misc.py:
```diff
@@ -172,10 +213,8 @@ def detect_win_roots(self):
-            if os.path.isdir(incdir):
-                return os.path.join(root, 'include')
+        if self.boost_root:
+            return os.path.join(self.boost_root,'include')
```
__[Flake8]__

[E231] missing whitespace after ','


<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/dependencies/misc.py:
```diff
@@ -958,3 +1013,117 @@ def get_methods(self):
 ]
+
+BOOST_DIRS = [
+#    'boost',
```
__[Flake8]__

[E122] continuation line missing indentation or outdented


<!-- Comment by [SideCI](https://sideci.com) -->

bredelings commented:
@mojca @NickeZ With these changes, I can build for either macports or homebrew with them both installed.  I'd like to get the boost test for homebrew working.  This change to `run_project_tests.py` enables the test:
```
-    elif mesonlib.is_osx() or mesonlib.is_cygwin():
+    elif mesonlib.is_osx():
+        # Just do the BOOST test
+        gathered_tests += [('framework', ['test cases/frameworks/1 boost'], False)]
+    elif mesonlib.is_cygwin():
+        # Skip all the framework tests
```
However, the test fails, saying that `#include <boost/version.hpp>` can't find the file.  On the mac I have access to, going into the boost dir and typing 'meson build' works though.

Any ideas?  I thought that homebrew boost was installed already on travis osx images (based on @NickeZ's boost-brew branch), so I would expect `/usr/local/include/boost/version.hpp` to exist.  I hear that on recent osx versions, you have to execute `xcode-select --install` in order to make the cli tools search /usr/local, so maybe we have to do this first?

bredelings commented:
@jpakkane Now that we can test boost on mac, is this OK?  If there is anything you'd like me to change about the approach please let me know.

jpakkane commented:
Looks good, but I'm not a fan of hardcoding the list of directories because it will get out of date. Could we instead get it from the system when running the checkup the first time? Something like:

    BOOST_DIRS=None
    ...
    if BOOST_DIRS is None:
        BOOST_DIRS = <glob from filesystem>
    ...
    <rest of code here>

bredelings commented:
@jpakkane My rationale is
* Since the warning is just to avoid failing on old `meson.build` files that list non-library modules, I don't think we need to update the list of boost dirs.  The code above says 'This will be an error in the future', so it seems like a deprecation warning.  Presumably after a year we can delete the BOOST_DIRS and the warning as well?
* On Unix we don't always know which directory contains the header files, so I don't know how to find the list of dirs from the filesystem in that case.  (If we could somehow get the list of directories that the compiler searches when we provide a list of -Idir flags, then we could probably search these dirs ourselves, but I don't know how to do this.)
Does that seem sensible?

This raises another problem that could perhaps be dealt with in the future.  I don't think we can in general give an error if people give a module that we don't know, because it could be a library from a future version of boost.  The library also might not be a directory name under /usr/include/boost/ even in the future version of boost -- consider `libboost_test_exec_monitor.a`.  What we can do is issue a warning that we don't recognize the module as a library, assume the module is a library anyway, and fail if we can't link against the library.  Thoughts?

jpakkane commented:
Let's merge this. We can fix issues later if they arise. Thanks.
